### PR TITLE
Versionize composer.lock

### PR DESCRIPTION
**Pull request by kingcrunch**
To ensure, that always the exactly same versions of the dependencies were installed,
the `composer.lock` should be versionized too. From time to time

```
php composer.phar update
```

should update the lock-file

### CODE REVIEW

**Comment by jayzeng** [comment]
Why not specify a version for  "ezyang/htmlpurifier": "dev-master"?

**Comment by kingcrunch** [comment]
@jazeng I have no idea :smile: This is out of my scope and because the tests passes with the current dev-version it seems to be no problem.

But as you mentioned HTMLPurifier: Could it be an issue, that it is not supporting (at least not regarding the docs) HTML5?

**Comment by jayzeng** [comment]
hmmm, that could be a different issue. I think this project should be positioned as a core library that takes care detection, with an adapter to adapt htmlpurifier or whatever users want to specify. But that's outside the scope of this pull, I personally feel it is weird to NOT ignore a file (composer.lock) that will be changed every time, and I can see it will be a source of merge conflict. Just my two cents :smile:

**Comment by kingcrunch** [comment]
@jayzeng `composer.lock` should not change every time, because one should _not_ call `update` over and over again. Only use `install`, except you definitely _want_ to update the dependencies, but this includes, that you test the application againts the updated packages and as soon as you have more than a handfull dependencies this is a task on its own :wink:

The `composer.lock` exists to ensure, that on every `install`-call the exactly same versions of dependencies were installed. The problem, if you don't versionize it, is that every user (that includes the developers and the CIs like travis) may have completely different versions including possible regression bugs and changed behaviour (especially when using a `-dev`-branch). This means, that you cannot rely on the CIs results anymore (the results always depend on _when_ the build runs) and you may get weird and unreproducable issues from users, that simply use broken versions of one or more dependecies. Oh, and of course it can break a live system.

TL;DR:
- Always use `install`, never use `update` in your normal workflow
- Except you _want_ to update the dependencies (which is ideally a separate and dedicated PR)
- If you want to add a new package, use `update vendor/packageName` to ensure, that `update` does not affect already registered (and tested) packages

At least this is the idea of the lock-file and why it should get versionized :smile: I for myself call `update` for my projects around once a week (or even less: as long as it works, there is no urge) and it works fine. As mentioned: This is a separate issue and commit.

**Comment by lstrojny** [comment]
The issue with htmlpurifier is, that there is currently no tag with composer support. That’s why dev-master is used. As soon as their is a tag, it wil be changed.

**Comment by kingcrunch** [comment]
@lstrojny And whats the reason the branch remains unmerged? :confused:

**Comment by lstrojny** [comment]
@KingCrunch sorry for not being clear enough. Versioning the composer.lock file isn’t something upstream projects (like PHPIDS) should do, but what local projects should do. The reason one could think we should do it for PHPIDS is, that htmlpurifier doesn’t have a tag with composer support right now and we need to use dev-master which might introduce breakage some time in the future. I'm living with that risk for now because htmlpurifier has a history of quite solidly managing compatibility, even in master.

**Comment by kingcrunch** [comment]
Well, even though I see your point, a don't see a downside in having the lock-file versionized, but I still don't feel comfortable in having an unpredictable installation on every checkout (which is the case especially for travis).

Thanks for the clarification :smile:
